plugins/authentication/src/native.cpp

```diff
@@ -291,13 +332,71 @@ namespace irods
             svr_req[irods_auth::next_operation] = AUTH_AGENT_AUTH_RESPONSE;
             auto resp = irods_auth::request(comm, svr_req);
 
-            comm.loggedIn = 1;
-
-            resp[irods_auth::next_operation] = irods_auth::flow_complete;
+            // If the client requested recording the auth file, that will be done here. Otherwise, authentication is
+            // complete.
+            if (const auto record_auth_file_iter = req.find(irods_auth::record_auth_file);
+                req.end() != record_auth_file_iter && record_auth_file_iter->get<bool>())
+            {
+                resp[irods::AUTH_PASSWORD_KEY] = client_provided_password_copy;
+                resp[irods_auth::next_operation] = record_authentication_file;
+            }
+            else {
+                resp[irods_auth::next_operation] = complete_authentication;
+            }
 
             return resp;
         } // native_auth_client_response
 
+        auto record_authentication_file_op(RcComm& _comm, const nlohmann::json& _req) -> nlohmann::json
+        {
+            // Get the password from the request payload, if available. If it is not there, this is not necessarily an
+            // error. In that case, just complete the authentication flow without recording an authentication file.
+            const std::string* password{};
+            const auto password_iter = _req.find(irods::AUTH_PASSWORD_KEY);
+            if (password_iter != _req.end()) {
+                password = password_iter->get_ptr<const std::string*>();
+            }
+            if (nullptr == password || password->empty()) {
+                return nlohmann::json{{irods_auth::next_operation, complete_authentication}};
+            }
+
+            // If TTL was provided, get and record the limited password.
+            if (const auto ttl_iter = _req.find(irods::AUTH_TTL_KEY); ttl_iter != _req.end()) {
+                if (const auto& ttl_str = ttl_iter->get_ref<const std::string&>(); !ttl_str.empty()) {
+                    try {
+                        // Specifically check for TTL of 0 here because 0 means "no TTL" for native authentication and
+                        // is a valid value as far as this authentication plugin is concerned. However, the mechanism
+                        // for getting the limited password considers non-positive values invalid. Therefore, we drop
+                        // through to the "else" case for 0 and pass non-zero values to the limited password machinery
+                        // to allow it to perform the value checking.
+                        if (const auto ttl = boost::lexical_cast<int>(ttl_str); 0 != ttl) {
```

Comment:
Why use `boost::lexical_cast` instead of `std::stoi`?

Reply:
Mostly because I copied this from elsewhere and tweaked it. Can switch to `std::stoi`.

Reply:
Switched